tools/cat2table.ml

```diff
@@ -0,0 +1,952 @@
+    case are sorted based on itself *)
+  let compare_expr e1 e2 =
+    let is_commutative = function
+    | Union | Inter | Add -> true
+    | _ -> false in
+
+    let do_compare_pairs (a1, b1) (a2, b2) cmp1 cmp2 =
+      let c = cmp1 a1 a2 in
+      if c != 0 then c
+      else cmp2 b1 b2 in
+
+    let rec do_compare_expr e1 e2 = match e1, e2 with
+    | e1, e2 when is_empty e1 && is_empty e2 -> 0
+    | Konst (_, (Universe _)), Konst (_, (Universe _)) -> 0
+    | e, _ when is_empty e -> -1
+    | _, e when is_empty e -> 1
+    | Konst (_, (Universe _)), _ -> -1
+    | _, Konst (_, (Universe _)) -> 1
+    | Var (_, id1), Var (_, id2) -> String.compare id1 id2
+    | Var _, _ -> -1
+    | _, Var _ -> 1
+    | Op1 (_, op1, e1), Op1 (_, op2, e2) ->
+      do_compare_pairs (op1, e1) (op2, e2) compare do_compare_expr
+    | Op1 _, _ -> -1
+    | _, Op1 _ -> 1
+    | Op (_, op1, es1), Op (_, op2, es2) ->
+      do_compare_pairs (op1, es1) (op2, es2) compare do_compare_expr_list
+    | Op _, _ -> -1
+    | _, Op _ -> 1
+    | App (_, e11, e12), App (_, e21, e22) ->
+      do_compare_expr_pairs (e11, e12) (e21, e22)
+    | App _, _ -> -1
+    | _, App _ -> 1
+    | Try (_, e11, e12), Try (_, e21, e22) ->
+      do_compare_expr_pairs (e11, e12) (e21, e22)
+    | Try _, _ -> -1
+    | _, Try _ -> 1
+    | If (_, c1, e11, e12), If (_, c2, e21, e22) ->
+      do_compare_pairs (c1, (e11, e12)) (c2, (e21, e22))
+        do_compare_cond
+        do_compare_expr_pairs
+    | If _, _ -> -1
+    | _, If _ -> 1
+    | _, _ -> 0 (* We don't care about the rest, since they are not effects *)
+
+    and do_compare_expr_pairs p1 p2 =
+      do_compare_pairs p1 p2 do_compare_expr do_compare_expr
+
+    and do_compare_cond c1 c2 = match c1, c2 with
+    | Eq (e11, e12), Eq (e21, e22) ->
+      do_compare_expr_pairs (e11, e12) (e21, e22)
+    | Eq _, _ -> -1
+    | _ , Eq _ -> 1
+    | Subset (e11, e12), Subset (e21, e22) -> do_compare_expr_pairs (e11, e12) (e21, e22)
+    | Subset _, _ -> -1
+    | _, Subset _ -> 1
+    | In (e11, e12), In (e21, e22) -> do_compare_expr_pairs (e11, e12) (e21, e22)
+    | In _, _ -> -1
+    | _, In _ -> 1
+    | VariantCond vc1, VariantCond vc2 -> do_compare_variant_cond vc1 vc2
+
+    and do_compare_variant_cond vc1 vc2 = match vc1, vc2 with
+    | Variant s1, Variant s2 -> String.compare s1 s2
+    | Variant _, _ -> -1
+    | _, Variant _ -> 1
+    | OpNot vc1, OpNot vc2 -> do_compare_variant_cond vc1 vc2
+    | OpNot _, _ -> -1
+    | _, OpNot _ -> 1
+    | OpAnd (vc11, vc12), OpAnd (vc21, vc22) ->
+      do_compare_pairs (vc11, vc12) (vc21, vc22)
+        do_compare_variant_cond do_compare_variant_cond
+    | OpAnd _, _ -> -1
+    | _, OpAnd _ -> 1
+    | OpOr (vc11, vc12), OpOr (vc21, vc22) ->
+      do_compare_pairs (vc11, vc12) (vc21, vc22)
+        do_compare_variant_cond do_compare_variant_cond
+
+    and do_compare_expr_list es1 es2 = match es1, es2 with
```

Comment:
I think it might be worth extracting this pattern into a generic utility function `compare : ('a -> 'a -> int) -> 'a list -> 'a list -> int`, and then instantiating it here as appropriate. In fact I would expect this `compare` pattern to be useful to other parts of the project too, so I would even put it in `Misc` or some other common location in `lib/`.